src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright (c) 2017, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file circuitpadding.h
+ * \brief Header file for circuitpadding.c.
+ **/
+#ifndef TOR_CIRCUITPADDING_H
+#define TOR_CIRCUITPADDING_H
+
+#include "circpad_negotiation.h"
+#include "lib/evloop/timers.h"
+
+typedef struct circuit_t circuit_t;
+typedef struct origin_circuit_t origin_circuit_t;
+typedef struct cell_t cell_t;
+
+/**
+ * Circpad state specifier.
+ *
+ * Each circuit has up to two state machines, and each state
+ * machine consists of these states. Machines transition between
+ * these states using the event transition specifiers below.
+ */
+typedef enum {
+  CIRCPAD_STATE_START = 0,
+  CIRCPAD_STATE_BURST = 1,
+  CIRCPAD_STATE_GAP = 2,
+  CIRCPAD_STATE_END = 3
+} circpad_statenum_t;
+#define CIRCPAD_NUM_STATES  ((uint8_t)CIRCPAD_STATE_END+1)
+
+/**
+ * These constants form a bitfield to specify the types of events
+ * that can cause transitions between state machine states.
+ *
+ * Note that SENT and RECV are relative to this endpoint. For
+ * relays, SENT means packets destined towards the client and
+ * RECV means packets destined towards the relay. On the client,
+ * SENT means packets destined towards the relay, where as RECV
+ * means packets destined towards the client.
+ */
+typedef enum {
```

Comment:
I'm a little confused why these are bits; it seems to me that a short array where you do something like `nextstate[eventno]` would be easier... but I haven't read the rest yet

Reply:
We do the inverse, actually. These are bits because there can be many more ways to get from one state to the other than there are states. So bitfield is a more compact way to represent these many transition arcs.

So instead of nextstate[event1] = next_state; nextstate[event2] = next_state; nextstate[event3] = next_state; we have transiton_events[next_state] |= event1|event2|event3.

Reply:
Ok, inverting this does make sense if we may have more states than events. Since I think we do want to have the option for more states, I did this inversion.

Header changes:  62748b4a3b7413d56ce68009c0003ffe7fe531ec. Implementation changes: bf45f1866702a7d3586a0c1a479aacb4a4ede55c. Test changes: c54dab42ca887445c9102871b30956eee71c9ae0